### PR TITLE
Test build and run Nix Flake in CI

### PR DESCRIPTION
**Pull request by alerque**
See discussion in #1450

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -35,3 +39,21 @@ jobs:
...
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - run: |
+          git fetch --prune --tags ||:
+      - uses: cachix/install-nix-action@v21
+        with:
+          extra_nix_config: |
+            access-tokens = github.com=${{ secrets.GITHUB_TOKEN }}
```
Comment (figsoda) [inline .github/workflows/ci.yml:55]: this is the default

```suggestion
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -35,3 +39,21 @@ jobs:
...
+      - run: nix flake check --impure
```
Comment (figsoda) [inline .github/workflows/ci.yml:56]:
```suggestion
      - run: nix flake check
```

**Comment by figsoda** [comment]
also I should probably note that the current approach doesn't have caching set up

File: .github/workflows/ci.yml (hunk 1 above)
Comment (alerque) [reply .github/workflows/ci.yml]: Thanks, I was unaware that had changed.

**Comment by alerque** [comment]
True. Also the other CI jobs in the same workflow don't have any Rust crate caching either.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (figsoda) [reply .github/workflows/ci.yml]: for reference this was added in v19

https://github.com/cachix/install-nix-action/commit/a4b8aaf5a24260b643f571235829e53e6b3cc920

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -35,3 +39,18 @@ jobs:
...
+
+  nix-flake:
+    name: Test build Nix Flake
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - run: |
+          git fetch --prune --tags ||:
```
Comment (figsoda) [inline .github/workflows/ci.yml:51]: what does this do?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -1,6 +1,10 @@
 name: Continuous integration
 on: [push, pull_request]
 
+concurrency:
+  group: ${{ github.workflow}}-${{ github.ref }}
```
Comment (figsoda) [inline .github/workflows/ci.yml:5]: just a styling nitpick

```suggestion
  group: ${{ github.workflow }}-${{ github.ref }}
```

**Review by figsoda — COMMENTED** [review]
lgtm

**Comment by laurmaedje** [comment]
> True. Also the other CI jobs in the same workflow don't have any Rust crate caching either.

They all use `Swatinem/rust-cache@v2`.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -1,6 +1,10 @@
 name: Continuous integration
 on: [push, pull_request]
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
```
Comment (laurmaedje) [inline .github/workflows/ci.yml:7]: Why this? Will this stop all workflows if one fails? If so, I would rather not have it as it's useful to know e.g. about clippy issues even if tests fail.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -35,3 +39,14 @@ jobs:
       - uses: dtolnay/rust-toolchain@1.70.0
       - uses: Swatinem/rust-cache@v2
       - run: cargo check --workspace
+
+  nix-flake:
+    name: Test build Nix Flake
```
Comment (laurmaedje) [inline .github/workflows/ci.yml:44]:
```suggestion
    name: Test Nix Flake
```

File: .github/workflows/ci.yml (hunk 5 above)
Comment (laurmaedje) [reply .github/workflows/ci.yml:7]: I see, then it makes sense.

**Comment by alerque** [comment]
> > True. Also the other CI jobs in the same workflow don't have any Rust crate caching either.
> 
> They all use `Swatinem/rust-cache@v2`.

My bad, apparently I can't read. Actually I'm just used to `uses:` lines that have 2 lines each with a name field, so I scanned too fast.

**Comment by figsoda** [comment]
I only have experience with cachix. Setting up sould be pretty easy, but there is not much I can do since I am not part of the typst team. There are a few more issues though:
- the package rebuilds even when irrelevant files change, we can filter `src` so it only rebuilds when the rust files change
- the compilation is not incremental, so the package needs to be rebuilt from ground up whenever anything changes. we can use [crate2nix](https://github.com/kolloch/crate2nix) but it might be a little too much set up, or [crane](https://github.com/ipetkov/crane) which caches the dependencies only, but it would be easier to set up

**Comment by alerque** [comment]
@fixsoda How about [magic-nix-cache](https://determinate.systems/posts/magic-nix-cache)?

**Comment by alerque** [comment]
I think we have a winner. No custom cache nonsense and no 3rd party service to sign up for, just a clever use of GH Actions built in cache system.

I ran the Nix test here twice. With a cold cache it was taking 5:28 to check the flake and similar for the build, with a warm cache it was 0:23 for the check and 0:04 for the build.

**Comment by figsoda** [comment]
Looks amazing! My points in https://github.com/typst/typst/pull/1477#issuecomment-1587576892 still applies, but I think this would be a good start.

**Comment by figsoda** [comment]
Opened #1586 to make sure the cache is a little more useful across updates

**Comment by laurmaedje** [comment]
Thanks!

**Comment by laurmaedje** [comment]
Looks like the nix flake caching has been, well, rather flakey recently. The CI actions very frequently take 6-9m with the tests only taking ~3m. Any ideas what is going on?

**Comment by alerque** [comment]
No but I've been on vacation for a couple weeks and haven't been following closely. I'll leave this notification around to look into it when I get a chance. If it falls off the radar feel free to open a new issue and assign it to me.

**Comment by laurmaedje** [comment]
@alerque I created an issue (https://github.com/typst/typst/issues/2047), but I can't assign you unless you leave a comment.

**Pull request merged**